src/AWS/Orleans.Clustering.DynamoDB/Options/DynamoDBClusteringOptions.cs

```diff
@@ -1,36 +1,9 @@
-﻿using Orleans.Clustering.DynamoDB;
+using Orleans.Clustering.DynamoDB;
 
 namespace Orleans.Configuration
 {
-    public class DynamoDBClusteringOptions
+    public class DynamoDBClusteringOptions : DynamoDBClientOptions
     {
-        /// <summary>
-        /// AccessKey string for DynamoDB Storage
-        /// </summary>
-        [Redact]
-        public string AccessKey { get; set; }
-
-        /// <summary>
-        /// Secret key for DynamoDB storage
-        /// </summary>
-        [Redact]
-        public string SecretKey { get; set; }
-
-        /// <summary>
-        /// DynamoDB Service name 
-        /// </summary>
-        public string Service { get; set; }
-
-        /// <summary>
-        /// Read capacity unit for DynamoDB storage
-        /// </summary>
-        public int ReadCapacityUnits { get; set; } = DynamoDBStorage.DefaultReadCapacityUnits;
-
-        /// <summary>
-        /// Write capacity unit for DynamoDB storage
-        /// </summary>
-        public int WriteCapacityUnits { get; set; } = DynamoDBStorage.DefaultWriteCapacityUnits;
```

Comment:
I'd rebase your branch on top of mine (https://github.com/dotnet/orleans/pull/7083) as:

* not all of these base properties are actually universally reused between all options
* not all of these base properties currently have default values
* that would simplify this PR

Reply:
> not all of these base properties are actually universally reused between all options

I don't think you're quite correct here, the implementation is the same copy of `DynamoDBStorage`, just with different [namespaces](https://github.com/dotnet/orleans/blob/main/src/AWS/Shared/Storage/DynamoDBStorage.cs#L13-L25). All of these implementations have a path that can call [Create Table](https://github.com/dotnet/orleans/blob/main/src/AWS/Shared/Storage/DynamoDBStorage.cs#L143-L147) which is called by the call to [`InitializeTable`](https://github.com/dotnet/orleans/search?q=InitializeTable)

> not all of these base properties currently have default values

I assume that you specifically mean the `WriteCapacityUnits` and `ReadCapacityUnits`? These have been copied directly from the old constant values [here](https://github.com/dotnet/orleans/pull/7019/files#diff-a0c73cd65ee261c035713e9d8e311ebd7ef4b3fac7898cca59a65d5892c40e6bL37-L38) to their new place on the options [here](https://github.com/dotnet/orleans/pull/7019/files#diff-f33823f0e1c63032342247595f3cc3d783b1205ca8993301367f2ab1bb7b3ec9R34-R41). We could keep the constant, but I personally don't see the value of it when it's defaulted onto the core options class. Given the internal nature of the class, I don't believe this to be a breaking change from an API surface perspective, but could see how it is a binary breaking change if we were using reflection to get these values.

> that would simplify this PR

Part of this PR is to ensure that the refactoring of the options is passed through to the underlying storage, which would mean that for your change [here](https://github.com/dotnet/orleans/pull/7082/files#diff-f4c3a22531bf7399cf4db98c6d65e2b53e9cbe60e27edd28be271234cd7b1e4dR43-R44) would be unnecessary as the underlying options would already have the appropriate changes made. Your change, after this PR, you would only need to add the associated option(s) to `DynamoDBClientOptions` & use the new options in `DynamoDbStorage` as required, which would take affect across all of the implementations in each of the namespaces.

Please let me know if you think there is something unclear here and I'll happily discuss :)

Reply:
> * not all of these base properties are actually universally reused between all options

That is technically true; however, I believe one of the benefits of consolidation is that options that have been added to only a subset of use cases are now available for all of them.

There is a mixing of concerns happening with the options classes, but that isn't new. Perhaps this would be an opportunity to split authentication options from table properties from functional concerns, but if we want to tackle that, it may be best to just jump to supporting injected clients / credentials.

Reply:
> not all of these base properties are actually universally reused between all options

No, I am correct. Properties like:

* `UseProvisionedThroughput`
* `EndpointDiscoveryEnabled`

are present in the proposed base class `DynamoDBClientOptions` https://github.com/dotnet/orleans/pull/7019/files#diff-f33823f0e1c63032342247595f3cc3d783b1205ca8993301367f2ab1bb7b3ec9 (which all dynamodb options are proposed to inherit from), but are not present in e.g. `DynamoDBClusteringOptions` https://github.com/dotnet/orleans/blob/main/src/AWS/Orleans.Clustering.DynamoDB/Options/DynamoDBClusteringOptions.cs

In my PR https://github.com/dotnet/orleans/pull/7083 I extract the common denominator for all propseries which are positively used in all types of dynamodb options. The extraction is very cautious, I could have gone with more properties like `TableName`, `ReadCapacityUnits` and `WriteCapacityUnits` (but ultimately decided not to as those have different default values), and I'm happy to add it if there's appetite for it in review comments.

> not all of these base properties currently have default values

I don't have a problem with constants, it seems to be a good move to me, it's just yet another thing in the already busy PR. If it was me, I'd introduce it in another set of changes to make this PR easier to review.

> that would simplify this PR

This is actually not clear to me. https://github.com/dotnet/orleans/pull/7082 is parked and waiting for your PR https://github.com/dotnet/orleans/pull/7019 to get completed, so I am going to rebase my changes on top of yours anyway. What I propose is that you reduce the complexity and the # of LoC of your PR by rebasing it on top of https://github.com/dotnet/orleans/pull/7083.

> There is a mixing of concerns happening with the options classes, but that isn't new. Perhaps this would be an opportunity to split authentication options from table properties from functional concerns, but if we want to tackle that, it may be best to just jump to supporting injected clients / credentials.

I think this is a great idea and is actually very simple to implement when https://github.com/dotnet/orleans/pull/7083 takes precedence over this PR as my PR only covers credentials-related properties. I would be quite happy to do it if there are no other takers.

Reply:
@PiotrJustyna apologizes for the delay. I was hoping someone from the Orleans team would chime in, but seeing as that hasn't happened, I'll give some more thoughts.

- I prefer this more encompassing refactoring. Ultimately all of these uses call into the same storage layer. Having unity in what is passed without a huge amount of constructor parameters seems like a clear win.
- It would be nice if authentication was pulled out slightly from the rest, but that does match existing implementation and the corresponding Azure options classes.
- I personally don't see this is too large a scope for a review, but I'll leave that decision to the Orleans team. If anything were to be pulled out, I'd say the additional new option.
- Tangential to this, the desired new option here and the new authentication options in the other PR could instead be handled by allowing DI for the dynamo client itself. Would need to default to one configured with the options if not available. That is definitely outside of this scope though.

Reply:
I’m not sure this is the PR you planned to comment on?

Reply:
> I’m not sure this is the PR you planned to comment on?

It is. I realize you asked about the other one, but this is where the discussion has been taking place.